### PR TITLE
ldms list type (heap)

### PR DESCRIPTION
**Pull request by narategithub**

### CODE REVIEW

File: ldms/src/sampler/examples/list_sampler/list_sampler.c
Comment (baallan) [inline ldms/src/sampler/examples/list_sampler/list_sampler.c:67]: why do we need ../../sampler_base.h? we seem to use unqualified in several other samplers.

**Comment by narategithub** [comment]
@tom95858 

This is currently a draft. The patches were ported from @tom `ovis-list.tgz`. It built, but I have not tested it yet.

With its current stage, it will break the compatibility with OVIS-4 HEAD. In order to preserve the compatibility, I think we need the following:

1. the `ldms_set_hdr->heap_sz` must use the `reserved` space, In other words:
```c
struct ldms_set_hdr {
	/* The unique metric set producer name */
	char producer_name[LDMS_PRODUCER_NAME_MAX];
	uint64_t meta_gn;	/* Meta-data generation number */
	struct ldms_version version;	/* LDMS version */
	uint8_t flags;	/* Set format flags */
	uint8_t pad1;	/* data pad */
	uint8_t pad2;	/* data pad */
	uint8_t pad3;	/* data pad */
	uint32_t card;		/* Size of dictionary */
	uint32_t meta_sz;	/* size of meta data in bytes */
	uint32_t data_sz;	/* size of metric values in bytes */
	uint32_t heap_sz;	/* size of heap in bytes */
	uint32_t uid;           /* UID */
	uint32_t gid;           /* GID */
	uint32_t perm;          /* permission */
	uint32_t array_card;    /* number of sets in the set array */
	uint32_t reserved[8];	/* area reserved for compatible core updates */
	uint32_t dict[OVIS_FLEX];/* The attr/metric dictionary */
};
```
should be 
```c
	/* The unique metric set producer name */
	char producer_name[LDMS_PRODUCER_NAME_MAX];
	uint64_t meta_gn;	/* Meta-data generation number */
	struct ldms_version version;	/* LDMS version */
	uint8_t flags;	/* Set format flags */
	uint8_t pad1;	/* data pad */
	uint8_t pad2;	/* data pad */
	uint8_t pad3;	/* data pad */
	uint32_t card;		/* Size of dictionary */
	uint32_t meta_sz;	/* size of meta data in bytes */
	uint32_t data_sz;	/* size of metric values in bytes */
	uint32_t uid;           /* UID */
	uint32_t gid;           /* GID */
	uint32_t perm;          /* permission */
	uint32_t array_card;    /* number of sets in the set array */
	uint32_t heap_sz;	/* size of heap in bytes */
	uint32_t reserved[7];	/* area reserved for compatible core updates */
	uint32_t dict[OVIS_FLEX];/* The attr/metric dictionary */
```
( In the earlier call with you, I was mistakenly looking at the pad1, pad2, pad3. That's why I said we have only 24 more bits for the `heap_sz`).


2. The `ldms_set_hdr->data_sz` should be `actual_data_sz + heap_sz`. This is because OVIS-4 uses `ldms_set_hdr->data_sz` to calculate the location of the set data (in set array). In the case of the current set data index being non-zero, if the `data_sz` does not include `heap_sz`, the OVIS-4 client (e.g. `ldms_ls`) will calculate wrong set data address if it lookup the set from this patch. See: https://github.com/narategithub/ovis/blob/2729bab9fb554fd1d72073a85dace2c8097b4e79/ldms/src/core/ldms_private.h#L182

REMARK: Since OVIS-4 HEAD doesn't know `LDMS_V_LIST` and `LDMS_V_LIST_ENTRY`, it is obvious that the OVIS-4 client won't be able to 100% work with the new ldmsd that uses `LDMS_V_LIST`. The OVIS-4 ldmsd collecting data from the new ldmsd would probably be OK ... depending on how storage plugins handle unknown datatype. OVIS-4 `ldms_ls` that query the new ldmsd that uses `LDMS_V_LIST` will print `"Unknown metric type"` for that particular metric.

@tom95858 what do you think? Any comments / suggestions? I'll proceed with 1 and 2 and test it while waiting for your input.

**Comment by tom95858** [comment]
@narategithub, yeah, that's why I was a bit confused. I thought I had made much space for future features. You're use of that 1st word for the heap size is the appropriate use of that reserved space.

**Comment by tom95858** [comment]
@narategithub I don't think it is a necessary backward compatibility claim that ldms_ls from an old LDMS will work with a new LDMS. I think it's that a new aggregator will continue to work with older LDMS samplers, i.e. the aggregators can be newer, but the samplers can be older.

But I think we should socialize this claim. We discussed this at the 2019 LDMSCONN, but that was a while ago.

**Comment by narategithub** [comment]
@tom95858 got it.

**Comment by baallan** [comment]
Within a single series like v4, ls should keep working from old version hosts to new aggregators.
in particular, a cluster on 4.x (well the admin nodes on it hosting 4.x L1s) should be able to ls 4.y L2s at the analysis servers.  At the very least, the user of 4.x -> 4.y ls should get back a well formed error message describing the problem, not a crash.

**Comment by narategithub** [comment]
@tom95858 This is ready for your review. It passed the compatibility test, the [ldms_list_test](https://github.com/narategithub/ldms-test/blob/ldms_list/ldms_list_test) in `narategithub/ldms-list` (will merge to `ovis-hpc/ldms-list` after this PR is merged), and other test cases in `ldms-list` -- except `setgroup` which is a known bug that shall be addressed in another PR.

**Comment by tom95858** [comment]
> Within a single series like v4, ls should keep working from old version hosts to new aggregators.

This is specifically not a goal per the compatibility guidelines we discussed at LDMSCONN

> in particular, a cluster on 4.x (well the admin nodes on it hosting 4.x L1s) should be able to ls 4.y L2s at the analysis servers. At the very least, the user of 4.x -> 4.y ls should get back a well formed error message describing the problem, not a crash.

It should certainly not crash. We should test this actually. @narategithub, would you please add a forward "does not crash" test to the auto-tests? Specifically 4.3.3 --> pull request.

File: ldms/src/core/ldms.h
Comment (tom95858) [inline ldms/src/core/ldms.h:1352]: May add "This presumes that the sample() function will update all list entry data at each invocation."

File: ldms/src/core/ldms.h
Comment (tom95858) [inline ldms/src/core/ldms.h:1357]: Please remove REMARK:

File: ldms/src/core/ldms.h
Comment (tom95858) [inline ldms/src/core/ldms.h:1366]: "turing" --> "turning"

File: ldms/src/core/ldms.h
Comment (tom95858) [inline ldms/src/core/ldms.h:1368]: Perhaps rename the parameter 'on' to 'on_n_off'

File: ldms/src/core/ldms_heap.h
Comment (tom95858) [inline ldms/src/core/ldms_heap.h:5]: The dates should be updated here to 2021

File: ldms/src/core/ldms_heap.h
Comment (tom95858) [inline ldms/src/core/ldms_heap.h:87]: should [0] be OVIS_FLEX or is this really *start?

File: ldms/src/core/ldms_heap.h
Comment (tom95858) [inline ldms/src/core/ldms_heap.h:93]: This parameter list is incomplete/incorrect.

File: ldms/src/core/ldms_heap.h
Comment (tom95858) [inline ldms/src/core/ldms_heap.h:102]: This is missing the _heap_ and _base_ parameters

File: ldms/src/core/ldms_heap.h
Comment (tom95858) [inline ldms/src/core/ldms_heap.h:115]: Again the doc string doesn't match the signature. Please just fix these up everywhere.

File: ldms/src/core/ldms_core.h
Comment (tom95858) [inline ldms/src/core/ldms_core.h:198]: If __data_sz__ includes both the static + heap data size, then I think __data_sz__ is the correct name. Otherwise, I think that __data_heap_sz__ is confusing.

File: ldms/src/core/ldms_core.h
Comment (tom95858) [inline ldms/src/core/ldms_core.h:201]: Did you determine that you didn't need to consume any of the reserved space? If so, please remove this field.

**Review by tom95858 — CHANGES_REQUESTED** [review]

File: ldms/src/core/ldms_core.h
Comment (narategithub) [reply ldms/src/core/ldms_core.h:201]: It is already removed. The `red` part is the removed stuff.

File: ldms/src/core/ldms_heap.h
Comment (narategithub) [reply ldms/src/core/ldms_heap.h:87]: It should be OVIS_FLEX. I'm addressing it.

File: ldms/src/core/ldms_heap.h
Comment (narategithub) [reply ldms/src/core/ldms_heap.h:115]: Got it. Thanks for catching these.

File: ldms/src/sampler/examples/list_sampler/list_sampler.c
Comment (narategithub) [reply ldms/src/sampler/examples/list_sampler/list_sampler.c:67]: no need. I did this to hush the warning from the autocompletion tool I used with Vim and forgot to change back. Changing it back now.

**Comment by narategithub** [comment]
> 
> 
> > Within a single series like v4, ls should keep working from old version hosts to new aggregators.
> 
> This is specifically not a goal per the compatibility guidelines we discussed at LDMSCONN
> 
> > in particular, a cluster on 4.x (well the admin nodes on it hosting 4.x L1s) should be able to ls 4.y L2s at the analysis servers. At the very least, the user of 4.x -> 4.y ls should get back a well formed error message describing the problem, not a crash.
> 
> It should certainly not crash. We should test this actually. @narategithub, would you please add a forward "does not crash" test to the auto-tests? Specifically 4.3.3 --> pull request.

Sure. I'm adding it with this PR.

**Comment by narategithub** [comment]
@tom95858 ready for you to re-review.

File: ldms/src/core/ldms.h
Comment (tom95858) [inline ldms/src/core/ldms.h:1951]: parameter names don't match the doc string

File: ldms/src/core/ldms.h
Comment (tom95858) [inline ldms/src/core/ldms.h:1344]: Missing parameter specification.

File: ldms/src/core/ldms.h
Comment (tom95858) [inline ldms/src/core/ldms.h:1959]: More doc string issues...

File: ldms/src/core/ldms.h
Comment (tom95858) [inline ldms/src/core/ldms.h:1966]: and here too

**Review by tom95858 — CHANGES_REQUESTED** [review]
Other than the doc stuff, I think it's ready to merge...thanks @narategithub

**Comment by narategithub** [comment]
> 
> 
> Other than the doc stuff, I think it's ready to merge...thanks @narategithub

@tom95858 oops. sorry about that. I'd just pushed the changes to those doc strings. Could you please take a look once more?

**Pull request merged**
